conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/TezosIndexer.scala

```diff
@@ -352,177 +352,119 @@ class TezosIndexer(
-          frozenBalance = optionalDelegate.flatMap(x => extractBalance(x.frozen_balance)),
-          stakingBalance = optionalDelegate.flatMap(x => extractBalance(x.staking_balance)),
-          delegatedBalance = optionalDelegate.flatMap(x => extractBalance(x.delegated_balance)),
-          deactivated = optionalDelegate.exists(_.deactivated)
-        )
+    ): List[Tables.BakersRow] = {
+
+      def findUpdateDelegate(baker: Tables.BakersRow) = delegates.get(PublicKeyHash(baker.pkh))
+
+      bakers.map(baker => baker -> findUpdateDelegate(baker)).collect {
+        case (baker, Some(delegate)) =>
+          baker.copy(
+            balance = extractBalance(delegate.balance),
+            frozenBalance = extractBalance(delegate.frozen_balance),
+            stakingBalance = extractBalance(delegate.staking_balance),
+            delegatedBalance = extractBalance(delegate.delegated_balance),
+            deactivated = delegate.deactivated
+          )
       }
+    }
 
     /** Extracts balance from PositiveBigNumber */
     def extractBalance(balance: PositiveBigNumber): Option[BigDecimal] = balance match {
       case PositiveDecimal(value) => Some(value)
       case InvalidPositiveDecimal(_) => None
     }
 
-    /** Processes blocks and fetches needed data to produce Governance rows*/
+    /** Processes blocks and fetches needed data to produce Governance rows.
+      * We assume that the input mapping will include all block entries relevant
+      * to a batch processing, that is, there could be block keys with a corresponding
+      * empty list as value.
+      * This way we can collapse the arguments to a single one, instead of requiring
+      * a separate one just to list the blocks of interest, and expecting
+      * rolls to only have non-empty values.
+      *
+      * @param rollsPerBlock rolls listings of delegates for the blocks to be processed
+      */
     def processBlocksForGovernance(
-        blocks: List[Block],
-        listings: List[(Block, List[Voting.BakerRolls])]
+        rollsPerBlock: Map[Block, List[Voting.BakerRolls]]
     ): Future[Unit] = {
-      import TezosDatabaseConversions._
-      import tech.cryptonomic.conseil.common.util.Conversion.Syntax._
-
-      //DANGER Will Robinson! we assume in the rest of the code that we're not handling the genesis block
-      val safeBlocks = blocks.filterNot(block => block.data.metadata == GenesisMetadata)
-      val safeListings = listings.filterNot { case (block, rolls) => block.data.metadata == GenesisMetadata }
+      val aggregateGovernanceData =
+        TezosGovernanceOperations.aggregateData(db) _
+
+      // DANGER Will Robinson!
+      // We assume in the rest of the code that we won't be handling the genesis metadata.
+      // The original arguments must not be re-used directly to ensure such guarantee
+      val (blocks, levelListings) = {
+        val nonGenesis = rollsPerBlock.filterNot { case (block, _) => block.data.metadata == GenesisMetadata }
```

Comment:
ok, I re-used the "listings" term as taken from the tezos api endpoint... (which I've found confusing at times, but I thought was conventionally accepted as a term)

Reply:
> Should we rename `blocks` to `nonGenesisBlocks` for clarity?

I'm not sure, you know, because I see it not as a defining characteristic, as much as a accidental necessity.
We can't extract the values we need for the genesis block, but it's not that it's intrinsically meaningful to discard the genesis from processing.
To me it looks mildly misleading to identify those as nonGenesisBlocks, it's just blocks we need to process, and we can skip the genesis because there's no data there to process.
WDYT?

Reply:
> > Should we rename `blocks` to `nonGenesisBlocks` for clarity?
> 
> I'm not sure, you know, because I see it not as a defining characteristic, as much as a accidental necessity.
> We can't extract the values we need for the genesis block, but it's not that it's intrinsically meaningful to discard the genesis from processing.
> To me it looks mildly misleading to identify those as nonGenesisBlocks, it's just blocks we need to process, and we can skip the genesis because there's no data there to process.
> WDYT?

Interestingly, `rollsPerBlock` will not have an entry for the genesis block since the corresponding RPC query comes back with no value! Not that I am saying it's not good to always validate your data. 

I agree with your comment otherwise.

Reply:
> ok, I re-used the "listings" term as taken from the tezos api endpoint... (which I've found confusing at times, but I thought was conventionally accepted as a term)

I think the Tezos node uses listings within some context so it's clear what the listings are referring to. I think the RPC route is something like '../governance/listings'. That context is not explicitly present here.

Reply:
> Interestingly, `rollsPerBlock` will not have an entry for the genesis block since the corresponding RPC query comes back with no value! Not that I am saying it's not good to always validate your data.
> 

I think the same, but as you say, I would not risk, considering we already had troubles with those genesis metadata values